File: plugins/camera/ios/CameraPlugin.swift
//
//  CameraPlugin.swift — lecodes-plugins/camera
//
//  The "camera" registerView plugin: WHEN and WITH WHAT to open it is app (TS) code —
//  the SDK's CameraView() wrapper rides the generic NativeView channel, no per-plugin
//  bridge methods.
//

import UIKit
import LeCodesSDK

public enum LeCodesCameraPlugin {

    public static let pluginId = "camera"

    public static func register(in engine: LeCodesEngine) {
        engine.registerView("camera") { [weak engine] params, _ in
            CameraViewInstance(params, engine)
        }
    }
}

/// The "camera" registerView factory product: wraps CameraView with the channel methods
/// the SDK's CameraView plugin calls — `takePhoto` → { systemId, name, size } and
/// `setFacingMode("front"|"back")`.
final class CameraViewInstance: NativeViewInstance {

    private let cameraView: CameraView
    private weak var engine: LeCodesEngine?

    var view: UIView { cameraView }

    init(_ params: Any?, _ engine: LeCodesEngine?) {
        let facing = (params as? [String: Any])?["facingMode"] as? String
        self.cameraView = CameraView(facingMode: facing == "front" ? .front : .back)
        self.engine = engine
    }

    func call(_ method: String, _ args: [Any], _ settle: ChannelSettle) {
        switch method {
        case "setFacingMode":
            cameraView.cameraSetFacingMode(args.first as? String == "front" ? .front : .back)
            settle.resolve()

        case "takePhoto":
            cameraView.takePhoto { [weak engine] image in
                guard let image, let jpg = image.jpegData(compressionQuality: 0.95) else {
                    settle.reject("Failed to take photo")
                    return
                }
                guard let engine else {
                    settle.reject("Engine is gone")
                    return
                }
                let systemId = engine.putBuffer(jpg)
                settle.resolve(["systemId": systemId, "name": "image.jpg", "size": jpg.count])
            }

        default:
            settle.reject("Unknown method: \(method)")
        }
    }
}

File: plugins/camera/android/src/main/java/io/letary/lecodes/plugins/camera/CameraPlugin.kt
//
//  CameraPlugin.kt — lecodes-plugins/camera
//
//  The "camera" registerView plugin: WHEN and WITH WHAT to open it is app (TS) code —
//  the SDK's CameraView() wrapper rides the generic NativeView channel, no per-plugin
//  bridge methods.
//

package io.letary.lecodes.plugins.camera

import android.content.Context
import io.letary.lecodes.LecodesEngine
import io.letary.lecodes.utils.FetchBuffers
import io.letary.lecodes.utils.NativeViews
import org.json.JSONObject

object CameraPlugin {

    const val PLUGIN_ID = "camera"

    fun register(engine: LecodesEngine, context: Context) {
        val appContext = context.applicationContext
        engine.registerView("camera") { params, _ ->
            val facing = if (params.optString("facingMode") == "front") CameraFacingMode.FRONT else CameraFacingMode.BACK
            val view = CameraView(appContext, facing)
            NativeViews.Instance(view, onCall = { call ->
                when (call.method) {
                    "setFacingMode" -> {
                        val mode = if (call.args.optString(0) == "front") CameraFacingMode.FRONT else CameraFacingMode.BACK
                        view.cameraSetFacingMode(mode)
                        call.resolve()
                    }
                    "takePhoto" -> view.takePhoto { bytes ->
                        if (bytes == null) {
                            call.reject("Failed to take photo")
                        } else {
                            val systemId = FetchBuffers.add(bytes)
                            call.resolve(JSONObject()
                                .put("systemId", systemId)
                                .put("name", "image.jpg")
                                .put("size", bytes.size))
                        }
                    }
                    else -> call.reject("Unknown method: ${call.method}")
                }
            })
        }
    }
}
